modules/widgets/src/pure-js/widget-impl.ts

```diff
@@ -0,0 +1,106 @@
+// deck.gl
+// SPDX-License-Identifier: MIT
+// Copyright (c) vis.gl contributors
+
+/* global document */
+import {_deepEqual, _applyStyles, _removeStyles} from '@deck.gl/core';
+import type {Deck, Widget, WidgetPlacement} from '@deck.gl/core';
+
+export type WidgetImplProps = {
+  id?: string;
+  /** Widget positioning within the view. Default 'top-left'. */
+  placement?: WidgetPlacement;
+  /** CSS inline style overrides. */
+  style?: Partial<CSSStyleDeclaration>;
+  /** Additional CSS class. */
+  className?: string;
+};
+export abstract class WidgetImpl<PropsT extends WidgetImplProps> implements Widget<PropsT> {
+  id: string;
+  props: Required<PropsT>;
+
+  deck?: Deck<any>;
+  element?: HTMLDivElement;
+
+  constructor(props: Required<PropsT>) {
+    this.id = props.id || 'widget';
+    this.props = props;
+  }
+
+  abstract onAdd({deck}: {deck: Deck<any>}): HTMLDivElement;
+  abstract onRemove(): void;
+
+  setProps(props: Partial<PropsT>) {
+    const oldProps = this.props;
+    const el = this.element;
+    if (el) {
+      if (oldProps.className !== props.className) {
+        if (oldProps.className) el.classList.remove(oldProps.className);
+        if (props.className) el.classList.add(props.className);
+      }
+
+      if (!_deepEqual(oldProps.style, props.style, 1)) {
+        _removeStyles(el, oldProps.style);
+        _applyStyles(el, props.style);
+      }
+    }
+
+    Object.assign(this.props, props);
+  }
+
+  _createElement(
+    props: {
+      type?: string;
+      style?: Partial<CSSStyleDeclaration>;
+      classNames?: string[];
+      events?: {[key: string]: EventListenerOrEventListenerObject};
+      children?: (HTMLElement | SVGSVGElement)[];
+    },
+    children?: (HTMLElement | SVGSVGElement)[]
+  ): HTMLElement {
+    const {type = 'div', style = {}, classNames = [], events = {}} = props;
+    children ||= props.children || [];
+    const el = document.createElement(type);
+    classNames.filter(Boolean).forEach(className => el.classList.add(className));
+    this._applyStyles(el, style);
+    children.forEach(child => el.appendChild(child));
+    Object.entries(events).forEach(([event, handler]) => el.addEventListener(event, handler));
+    return el;
+  }
+
+  _createIconButton(props: {
```

Comment:
Everything in this base class seems like it could also be used by the preact widgets too, except for this `_createIconButton`, which would just use the preact-equivalent instead.

Reply:
> Everything in this base class seems like it could also be used by the preact widgets too,

Yes probably true. I cleaned it up further, removing even more boilerplate code from the widget.